File: src/medium/MinJumpToEnd.java
package src.medium;

/**
 * https://leetcode.com/problems/jump-game-ii
 */
public class MinJumpToEnd {

    public static void main(String[] args) {
        int[] nums = {2, 3, 1};
        System.out.println(jump(nums));
    }

    private static int jump(int[] nums) {
        int result = 0;
        int currEnd = 0;
        int nextIndex = 0;

        for (int i = 0; i < nums.length - 1; i++) {
            System.out.println("i = " + i);
            nextIndex = Math.max(nextIndex, i + nums[i]);
            if (i == currEnd) {
                result++;
                currEnd = nextIndex;
            }
        }

        return result;
    }
}

File: src/scala/MinJumpToEnd.scala
package src.scala

/**
 * @author Prashant Sharma
 *         https://leetcode.com/problems/jump-game-ii
 */
object MinJumpToEnd {

    def main(args: Array[String]): Unit = {
        val nums = Array(2, 3, 1)
        println(jump(nums))
    }

    def jump(nums: Array[Int]): Int = {
        var furthest = nums(0)
        var maxPos = 0
        var steps = 0

        1 until nums.length foreach (i => {
            if (i > maxPos) {
                steps += 1
                maxPos = furthest
            }
            if (maxPos >= nums.length - 1) {
                steps
            }
            else furthest = math.max(furthest, nums(i) + i)
        })
        steps
    }
}
